### PR TITLE
 BTREE.CPP

### PR DESCRIPTION
**Pull request by vipinmankar**
//B tree cpp program
#include <iostream>
using namespace std;
class btnode{
    public :
    int *value;
    int t;
    btnode **child;
    int n;
    bool leaf;
    
    btnode(int t1,bool leaf1){
        t=t1;
        leaf=leaf1;
        value=new int[2*t-1];
        child=new btnode *[2*t];
        n=0;
    }
    void split(int i,btnode *y){
        btnode *z=new btnode(y->t,y->leaf);
        z->n=t-1;
        for (int j = 0; j < t-1; j++)
            z->value[j] = y->value[j+t];
        
        
        if (y->leaf == false)
        {
            for (int j = 0; j < t; j++)
                z->child[j] = y->child[j+t];
        }
        
        
        y->n = t - 1;
        
        
        for (int j = n; j >= i+1; j--)
            child[j+1] = child[j];
        
        child[i+1] = z;
        
        for (int j = n-1; j >= i; j--)
            value[j+1] = value[j];
        
        value[i] = y->value[t-1];
        
        n = n + 1;  }
    void insertnonfull(int k){
        int i = n-1;
        
        if (leaf == true)
        {
            while (i >= 0 && value[i] > k)
            {
                value[i+1] = value[i];
                i--;
            }
            
            value[i+1] = k;
            n = n+1;
        }
        else
        {
            while (i >= 0 && value[i] > k)
                i--;
            
            if (child[i+1]->n == 2*t-1)
            {
                split(i+1, child[i+1]);
                
                if (value[i+1] < k)
                    i++;
            }
            
        }
        
    }
    void traverse(){
        int i;
        for (i = 0; i < n; i++)
        {
            if (leaf == false)
                child[i]->traverse();
            cout <<"\n"<< value[i];
        }
        
        
        if (leaf == false)
            child[i]->traverse();
    }
    friend class btree;
};
class btree{
public:
    btnode  *root;
    int t,temp;
    btree(int t1){
        root=NULL;
        t=t1;
    }
    void traverse()
    {  if (root != NULL) root->traverse(); }
    void insert(){
        cout<<" "<<"Enter the number to add"<<"\n" ;
        cin>>temp;
        if(root==NULL){
            root=new btnode(t,true);
            root->value[0]=temp;
            root->n=1;
            
        }
        else{
            if(root->n==(2*t-1)){
                btnode *s=new btnode(t,false);
                s->child[0]=root;
                s->split(0,root);
                int i=0;
                if (s->value[0] < temp)
                    i++;
                s->child[i]->insertnonfull( temp);
                root = s;
                
            }
            
            else{
                root->insertnonfull( temp);
            }
        }
    }
    
};

int main()
{
    btree t(3);
    int opt;
    while(1){
        cout<<" \npress 1 for insert \npress 2 for traverse \npress 3 for Exit ";
        cin>>opt;
        switch(opt){
            case 1:
                t.insert();
                break;
            case 2:
                t.traverse();
                break;
            case 3:
                exit(0);
        }
    }
    return 0;
}

//Sample input :
10 21 31 41 51 
output 
10 21 31 41 51
 //Sample input :
70 31 31 21 51 
output 
21 21 31 51 70

### CODE REVIEW

File: BTree.cpp [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+                t.traverse();
+                break;
+            case 3:
+                exit(0);
+        }
+    }
+    return 0;
+}
+//Sample input :
+10 21 31 41 51
```
Comment (raghav-dalmia) [inline BTree.cpp]: Comment these input/ output

File: BTree.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+class btree{
+public:
+    btnode  *root;
+    int t,temp;
+    btree(int t1){
+        root=NULL;
+        t=t1;
+    }
+    void traverse()
+    {  if (root != NULL) root->traverse(); }
```
Comment (raghav-dalmia) [inline BTree.cpp]: use proper indentation

File: BTree.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    void traverse(){
+        int i;
+        for (i = 0; i < n; i++)
+        {
+            if (leaf == false)
+                child[i]->traverse();
+            cout <<"\n"<< value[i];
+        }
+
```
Comment (raghav-dalmia) [inline BTree.cpp]: remove extra lines

File: BTree.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    
+    btnode(int t1,bool leaf1){
+        t=t1;
+        leaf=leaf1;
+        value=new int[2*t-1];
+        child=new btnode *[2*t];
+        n=0;
+    }
+    void split(int i,btnode *y){
+        btnode *z=new btnode(y->t,y->leaf);
```
Comment (raghav-dalmia) [inline BTree.cpp]: use proper variable names

**Comment by raghav-dalmia** [comment]
use the template for PR description, squash commits, add comments, give a small description about code at the beginning of code.

**Comment by vipinmankar** [comment]
Thank you for viewing my work and for your valuable suggestions.
I will try to complete the  assigned work at the earliest.

File: BTree.cpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 
 
 // cpp program of bTree  ton insert and traverse the tree ,BTree is tree data structure which use to
- //indexing of database.
```
Comment (vipinmankar) [inline BTree.cpp]: i

**Comment by vipinmankar** [comment]
@raghav-dalmia i have made the changes as you mention please review it.
Thank you

**Comment by vipinmankar** [comment]
@raghav-dalmia is there is any chances i have to make please let me know .

**Comment by jainaman224** [comment]
Fix issue template. It is compulsory.

File: BTree.cpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
+
+
+// cpp program of BTree on insert and traverse the tree ,BTree is tree data structure which use to
+ //indexing of database.//
+
+#include <iostream>
+using namespace std;
+class create_tree_node{
```
Comment (jainaman224) [inline BTree.cpp]: Add new line above this.

File: BTree.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    bool leaf;
+    
+    create_tree_node(int t1,bool leaf1){
+        no_of_key=t1;
+        leaf=leaf1;
+        Key=new int[2*no_of_key-1];
+        child=new create_tree_node *[2*no_of_key];
+        count_of_keys=0;
+    }
+    //Function to split note and creayte new node
```
Comment (jainaman224) [inline BTree.cpp]: Add space after `//`.

File: BTree.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    void split(int i,create_tree_node *y){
+        create_tree_node *z=new create_tree_node(y->no_of_key,y->leaf);   // create a new node pointer
+        z->count_of_keys=no_of_key-1;   //new node t-1 node
+       
+        // Copy the last (t-1) keys of y to z
+        for (int j = 0; j < no_of_key-1; j++)
+            z->Key[j] = y->Key[j+no_of_key];
+        
+         // Copy the last t children of y to z
```
Comment (jainaman224) [inline BTree.cpp]: Inconsistent indentation.

File: BTree.cpp [hunk 9]
```diff
@@ -0,0 +1,184 @@
+
```
Comment (jainaman224) [inline BTree.cpp]: Add this file to a folder.

**Comment by vipinmankar** [comment]
@raghav-dalmia sir @jainaman224 sir i have made the required changes please merge this PR.

File: BTree.cpp (hunk 6 above)
Comment (somya-kapoor) [inline BTree.cpp]: Leave a line before it

File: BTree.cpp [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+using namespace std;
+class create_tree_node{
+    public :
+    int *Key;
+    int no_of_key;
+    create_tree_node **child;
+    int count_of_keys;
+    bool leaf;
+    
+    create_tree_node(int t1,bool leaf1){
```
Comment (somya-kapoor) [inline BTree.cpp]: Leave a space after ,

File: BTree.cpp [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+        no_of_key=t1;
```
Comment (somya-kapoor) [inline BTree.cpp]: Add space around the operator

File: BTree.cpp [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+        leaf=leaf1;
```
Comment (somya-kapoor) [inline BTree.cpp]: Add space around the operator. Do this in whole file

File: BTree.cpp (hunk 7 above)
Comment (somya-kapoor) [inline BTree.cpp]: Add space after //.

File: BTree.cpp (hunk 8 above)
Comment (somya-kapoor) [inline BTree.cpp]: Inconsistent indentation.

File: BTree.cpp (hunk 9 above)
Comment (somya-kapoor) [inline BTree.cpp]: There are 2 files with the same code, delete one of them.

**Comment by somya-kapoor** [comment]
@vipinmankar Are you working on it?

File: BTree.cpp (hunk 9 above)
Comment (vipinmankar) [reply BTree.cpp:1]: @somya-kapoor yes maam , by tomorrow morning i will commit it with all required  changes.

**Comment by somya-kapoor** [comment]
@vipinmankar There is no file in the PR

**Comment by vipinmankar** [comment]
@somya-kapoor  i have added the file as it was not there  ,as you mention earlier that my repo contain 2 file of same name so after that  it i had deleted those from master branch and create a new branch made the new  pr but it  was not showing pr .So now i have again added it to the master and it can be now be viewed in pr. Hence I request you to review it . 
Thank you

**Comment by somya-kapoor** [comment]
@vipinmankar You have created 2 PR for the same issue #1894 and #2686. Please close on of them.

**Comment by somya-kapoor** [comment]
@vipinmankar Are you working on it?

**Pull request closed without merging**